### PR TITLE
Bugfix FXIOS-9998 Fix crash when deleting and the undoing the inactive tabs multiple times in a row (related improvements FXIOS-9954, FXIOS-10010, FXIOS-9999)

### PR DESCRIPTION
**Pull request by ih-codes**
## :scroll: Tickets
[Jira ticket FXIOS-9998](https://mozilla-hub.atlassian.net/browse/FXIOS-9998)
[Github issue 21953](https://github.com/mozilla-mobile/firefox-ios/issues/21953)

This will also fix (or improve at least) some other lingering inactive tab bugs:
[Jira ticket FXIOS-9954](https://mozilla-hub.atlassian.net/browse/FXIOS-9954) -- Improvement but not 100% fix
[Github issue 21863](https://github.com/mozilla-mobile/firefox-ios/issues/21863)

[Jira ticket FXIOS-9954](https://mozilla-hub.atlassian.net/browse/FXIOS-10010)
[Github issue 21971](https://github.com/mozilla-mobile/firefox-ios/issues/21971)

[Jira ticket FXIOS-9954](https://mozilla-hub.atlassian.net/browse/FXIOS-9999)
[Github issue 21954](https://github.com/mozilla-mobile/firefox-ios/issues/21954)

## :bulb: Description

I couldn't reproduce the crash in [FXIOS-9998](https://mozilla-hub.atlassian.net/browse/FXIOS-9998), but after inspecting the crash log it was clear there was something odd going on with wiping and restoring multiple inactive tabs.

So I cleaned up a lot of this logic and reduced dependencies to some of the inner methods (e.g. we don't need to calculate the "viable" deletion index prior to removing a tab, hold onto that index, and then pass it back into `updateSelectedTabAfterRemovalOf` later. We can just do some math in `findRightOrLeftTab`.)

Regrettably, the logic for `updateSelectedTabAfterRemovalOf` is more complicated now, but that's unavoidable as inactive tabs just weren't fully accounted for prior. I tried to document the complexity as best I could.

I also added a huge swath of unit tests because of this complex logic. If anyone has suggestions on making these prettier while still retaining their usefulness, my ears are open. The tests are a bit long so the bot might shame me for PR length. 😅 

Actual code changes to review are much smaller!!

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [x] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

[FXIOS-9998]: https://mozilla-hub.atlassian.net/browse/FXIOS-9998?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 1]
```diff
@@ -546,12 +555,9 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
...
         guard let index = tabs.firstIndex(where: { $0 === tab }) else { return }
-        DispatchQueue.main.async { [unowned self] in
-            // gather the index of the deleted tab within the viable tabs array
-            // so we can select the correct next tab after deletion
-            let viableTabsIndex = deletedIndexForViableTabs(tab)
-            self.removeTab(tab, flushToDisk: true)
-            self.updateIndexAfterRemovalOf(tab, deletedIndex: index, viableTabsIndex: viableTabsIndex)
+        DispatchQueue.main.async { [weak self] in
+            self?.removeTab(tab, flushToDisk: true)
+            self?.updateSelectedTabAfterRemovalOf(tab, deletedIndex: index)
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:560]: I was able to simplify things a bit by calculating this index inside `findRightOrLeftTab` instead of holding onto it and passing it in later.

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 2]
```diff
@@ -568,15 +574,14 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
...
         if TabTrayFlagManager.isRefactorEnabled {
             backupCloseTab = BackupCloseTab(
                 tab: tab,
-                restorePosition: viableTabsIndex,
+                restorePosition: index,
                 isSelected: selectedTab?.tabUUID == tab.tabUUID)
         }
         self.removeTab(tab, flushToDisk: true)
-        self.updateIndexAfterRemovalOf(tab, deletedIndex: index, viableTabsIndex: viableTabsIndex)
+        self.updateSelectedTabAfterRemovalOf(tab, deletedIndex: index) // Crash path
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:584]: Same simplification here for "viableTabsIndex"

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 3]
```diff
@@ -819,56 +824,96 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
         privateConfiguration.preferences.javaScriptCanOpenWindowsAutomatically = allowPopups
     }
 
-    // returns all activate tabs (private or normal)
-    private func viableTabs(isPrivate: Bool = false) -> [Tab] {
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:823]: This method didn't really seem useful anymore. There's a separate task to clean up the inactiveTabs refactor.

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 4]
```diff
@@ -905,24 +950,6 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
         storeChanges()
     }
 
-    // Select the most recently visited tab, IFF it is also the parent tab of the closed tab.
-    private func selectParentTab(afterRemoving tab: Tab) -> Bool {
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:909]: This was only used in one place, so this logic is essentially checked in `updateSelectedTabAfterRemovalOf` (formerly called `updateIndexAfterRemovalOf`).

File: firefox-ios/Client/TabManagement/TabManager.swift [hunk 5]
```diff
@@ -21,11 +21,12 @@ protocol TabManager: AnyObject {
...
     var count: Int { get }
     var selectedTab: Tab? { get }
+    var selectedTabUUID: UUID? { get }
     var backupCloseTab: BackupCloseTab? { get set }
     var backupCloseTabs: [Tab] { get set }
-    var normalTabs: [Tab] { get }
+    var normalTabs: [Tab] { get } // Includes active and inactive tabs
     var normalActiveTabs: [Tab] { get }
-    var inactiveTabs: [Tab] { get }
+    var normalInactiveTabs: [Tab] { get }
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/TabManager.swift:29]: Renamed for clarity since private tabs can't be inactive technically, or at least not as we are using "inactive". :)

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift [hunk 6]
```diff
@@ -189,16 +189,16 @@ class TabManagerImplementation: LegacyTabManager, Notifiable, WindowSimpleTabsPr
                    level: .debug,
                    category: .tabs)
 
-        selectTab(tabToSelect)
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/TabManagerImplementation.swift:192]: There was an edge case where tabToSelect might be nil so I thought we should handle that explicitly.

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift [hunk 7]
```diff
@@ -236,11 +236,9 @@ class TabManagerImplementation: LegacyTabManager, Notifiable, WindowSimpleTabsPr
 
     private func preserveTabs(forced: Bool) {
         Task {
-            // This value should never be nil but we need to still treat it
-            // as if it can be nil until the old code is removed
-            let activeTabID = UUID(uuidString: self.selectedTab?.tabUUID ?? "") ?? UUID()
+            // FIXME FXIOS-10059 TabManagerImplementation's preserveTabs is called with a nil selectedTab
             let windowData = WindowData(id: windowUUID,
-                                        activeTabId: activeTabID,
+                                        activeTabId: self.selectedTabUUID ?? UUID(),
```
Comment (ih-codes) [inline firefox-ios/Client/TabManagement/TabManagerImplementation.swift:241]: @OrlaM This is the spot where we see selectedTab as nil each time. Just turn on inactive tabs and close a bunch at once, it's 100% reproducible. I've filed a separate ticket as I do not think this is related to the crash I was investigating.

File: firefox-ios/Client/TabManagement/TabManager.swift (hunk 5 above)
Comment (dataports) [reply firefox-ios/Client/TabManagement/TabManager.swift:29]: Kind of a nit, but I vote keeping this to `inactiveTabs` because there's a lot of references to "inactiveTabs" without the normal modifier in the codebase that it might actually just be more mental overhead to change the name at this point. Maybe a comment here stating that private tabs can't be inactive would be better?

Comment (ih-codes) [reply firefox-ios/Client/TabManagement/TabManager.swift:29]: I can revert this pretty easily for now if we want to have a larger discussion on the value of this change!

I agree, other naming should be updated as well in this case. It could be a separate task if we want to go forward with a change like this.

For me, it was the lack of symmetry between `normalActiveTabs` and `inactiveTabs` that was a bit confusing. And that an "active tab" is assumed normal and not private, while an "inactive tab" is assumed also normal and not private. But a "private tab" can't be "inactive tab"—except actually, it can be, we just don't differentiate in the UI or have special private inactive tab logic haha. It starts to get a little confusing to know exactly what each subarray contains as there are now 2 flags at play. 😂 

While I was writing the tests I had to be very explicit with this, which was why I thought the naming could maybe be improved across the three tab types we really care about: normal active tabs, normal inactive tabs, and private tabs (technically either active or inactive).

Comment (dataports) [reply firefox-ios/Client/TabManagement/TabManager.swift:29]: Yeah maybe revert for now and lets have a larger discussion - I agree that something is not quite right with the naming for the tabs. Maybe we can discuss in the iOS devs channel and make a ticket to change the naming from that? You could basically just put what you wrote here.
If we're gonna fix it, we might as well fix it everywhere in the project 😆

File: firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift [hunk 8]
```diff
@@ -436,6 +439,14 @@ class TabManagerMiddleware {
...
             store.dispatch(refreshAction)
 
+            // Refresh the active tabs panel. Can only happen if the user is in normal browsering mode (not private).
+            // Related: FXIOS-10010, FXIOS-9954, FXIOS-9999
+            let model = getTabsDisplayModel(for: false, shouldScrollToTab: false, uuid: uuid)
+            let refreshActiveTabsPanelAction = TabPanelMiddlewareAction(tabDisplayModel: model,
+                                                                        windowUUID: uuid,
+                                                                        actionType: TabPanelMiddlewareActionType.refreshTabs)
+            store.dispatch(refreshActiveTabsPanelAction)
+
```
Comment (dataports) [inline firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift:449]: https://github.com/user-attachments/assets/896ff876-7344-4d3f-bfba-f7fb1fd6cffe

Not sure if this is just a weird thing with the debug menu 10s timeout for tab activity being enabled but it looks like this doesn't work anymore?

File: firefox-ios/Client/TabManagement/TabManager.swift (hunk 5 above)
Comment (ih-codes) [reply firefox-ios/Client/TabManagement/TabManager.swift:29]: Sounds good to me! I've started a discussion in slack and made a [proposal task](https://mozilla-hub.atlassian.net/browse/FXIOS-10089) for us to discuss! 👍  Will revert this change in the meantime.

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 9]
```diff
@@ -819,56 +824,96 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
...
+        // Once a tab has been removed, we must consider several situations:
+        // A) A change in the `tabs` array size, which may require us to shift the selected tab index 1 to the left
+        // B) If the removed tab was selected, we must choose an appropriate next selected tab from among the existing tabs
+        // C) When we close the last tab of a certain type, we may need to perform additional logic
+        //      - For example, closing the last private tab should select the most recent active normal tab, if possible
+
+        // Viable tabs for selection are private tabs and normal active tabs. We never want to select a normal inactive tab.
+        let viableTabsToSelect = removedTab.isPrivate
+                          ? privateTabs
+                          : normalActiveTabs
```
Comment (dataports) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:840]: I really like this spacing for ternary operator syntax btw, much easier to parse for me

**Comment by ih-codes** [comment]
Working on some bugfixes on this branch... don't bother reviewing until I work through them tomorrow! 🙏  Will close the PR just so no one wastes time on this until it's ready.

File: firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift (hunk 8 above)
Comment (ih-codes) [reply firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift:449]: Really great catch @dataports, I've updated my logic and added additional tests! 🎉 

I _think_ all the edge cases have now been accounted for.


https://github.com/user-attachments/assets/0c7db232-66c2-486b-9f77-babb82289d13

Comment (ih-codes) [reply firefox-ios/Client/Frontend/Browser/Tabs/Middleware/TabManagerMiddleware.swift:449]: (P.S. still hold off on reviewing, one other little thing I'll tweak tomorrow)

**Comment by ih-codes** [comment]
Cleaned this up a bit, fixed the edge case @dataports spotted, and added more tests to catch that situation. I think this is ready for review again. 🙏

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 10]
```diff
@@ -132,6 +131,15 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
         return tabs[_selectedIndex]
     }
 
+    var selectedTabUUID: UUID? {
+        guard let selectedTab = self.selectedTab,
+              let uuid = UUID(uuidString: selectedTab.tabUUID) else {
```
[the remark this answers is not in the document]
Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:136]: Yeah, let's have a larger discussion. I know Roux has voiced preference to this stylistic choice. I personally find it harder to read when the return paths are not all left aligned. 😅

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift (hunk 7 above)
Comment (ih-codes) [reply firefox-ios/Client/TabManagement/TabManagerImplementation.swift:241]: Forgot to link the ticket: https://mozilla-hub.atlassian.net/browse/FXIOS-10059

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 11]
```diff
@@ -819,56 +824,101 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
...
+                // 3. If there are no suitable active tabs to select, create a new normal active tab.
+                // (Note: It's possible to fall into here when all tabs have become inactive, especially when debugging.)
+                selectTab(addTab(), previous: removedTab)
             }
         } else if deletedIndex < _selectedIndex {
-            // If we delete an active tab that's before the selected active tab, we need to shift our selection index
-            // since the array size has changed.
-            let selected = tabs[safe: _selectedIndex - 1]
-            selectTab(selected, previous: selected)
+            // If we delete a tab in the `tabs` array that's earlier than the selected tab, we need to shift our index.
```
Comment (dataports) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:877]: "earlier than" -> "ahead of" maybe?  Unless I'm misunderstanding the comment

Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift]: Sure, that works! 👍 Committed.

**Review by dataports — APPROVED** [review]

**Review by clarmso — APPROVED** [review]
